### ISSUE TITLE
Not properly constructed object may be visible to user code

### ISSUE THREAD

**Issue by bigwad** [Starcounter/Starcounter#3215]
Please, consider following code:
```
[Database]
class Entity
{
   public Entity()
   {
       if (something)       
          throw exception;
       <initialization goes here>
   }
}

try
{
  new Entity();
}
catch( Exception )
{
  //starting from here we may select invalid entity object. we may even commit it
}
```

Constructors are to ensure object's invariant. So if we throw an exception from constructor - object shouldn't be created. Thus it's not right that user code may find an object with invalid invariant in database. I think, we should remove appropriate db record automatically if constructor throws an exception.

**Comment by per-samuelsson** [comment]
Yeah, this is probably somewhat a marginal case, but it seem right when I consider it, and if so, that's a bug. I'm not sure exactly how we should solve it, but lets try. I think we can just catch any exception in the weaved "insert constructor" and issue a delete if we catch any.

Good catch @bigwad!

I'll assign it to me to first validate it, but it will be after christmas I think.
